### ISSUE TITLE
CMAKE_BUILD_TYPE=Release but it says 'ikos was compiled in debug mode'

### ISSUE THREAD

**Issue by yurivict** [NASA-SW-VnV/ikos#36]
```
[!] ikos was compiled in debug mode, the analysis might be slow
```

**Comment by alex-justes** [comment]
How are you building it? 

I've got the same result only when i explicitly call `cmake .. -DCMAKE_BUILD_TYPE=Debug && cmake --build .`, otherwise NDEBUG flag is set, so that line is removed by the compiler. 

Try to use plain `cmake .. && cmake --build .`

P.S. 
Tried that on the 081999b6f2f0afd1fab02eed661f01ddd750c2fd commit

**Comment by yurivict** [comment]
I'm building it using the FreeBSD port framework. It always supplies ```-DCMAKE_BUILD_TYPE:STRING="Release"```:
```
-DLLVM_CONFIG_EXECUTABLE:PATH=/usr/local/bin/llvm-config70  -DPYTHON_EXECUTABLE:STRING=/usr/local/bin/python2.7 -DCMAKE_C_COMPILER:STRING="cc"  -DCMAKE_CXX_COMPILER:STRING="c++"  -DCMAKE_C_FLAGS:STRING="-O2 -pipe -fno-omit-frame-pointer  -fstack-protector -fno-strict-aliasing "  -DCMAKE_C_FLAGS_DEBUG:STRING="-O2 -pipe -fno-omit-frame-pointer  -fstack-protector -fno-strict-aliasing "  -DCMAKE_C_FLAGS_RELEASE:STRING="-O2 -pipe -fno-omit-frame-pointer  -fstack-protector -fno-strict-aliasing "  -DCMAKE_CXX_FLAGS:STRING="-O2 -pipe -fno-omit-frame-pointer -fstack-protector -fno-strict-aliasing -fno-omit-frame-pointer  "  -DCMAKE_CXX_FLAGS_DEBUG:STRING="-O2 -pipe -fno-omit-frame-pointer -fstack-protector -fno-strict-aliasing -fno-omit-frame-pointer  "  -DCMAKE_CXX_FLAGS_RELEASE:STRING="-O2 -pipe -fno-omit-frame-pointer -fstack-protector -fno-strict-aliasing -fno-omit-frame-pointer  "  -DCMAKE_EXE_LINKER_FLAGS:STRING=" -fstack-protector "  -DCMAKE_MODULE_LINKER_FLAGS:STRING=" -fstack-protector "  -DCMAKE_SHARED_LINKER_FLAGS:STRING=" -fstack-protector "  -DCMAKE_INSTALL_PREFIX:PATH="/usr/local"  -DCMAKE_BUILD_TYPE:STRING="Release"  -DTHREADS_HAVE_PTHREAD_ARG:BOOL=YES  -DCMAKE_INSTALL_RPATH_USE_LINK_PATH=YES  -DCMAKE_VERBOSE_MAKEFILE:BOOL=ON -GNinja -DPython_ADDITIONAL_VERSIONS=2.7 -DBOOST_PYTHON_SUFFIX:STRING=27
```

**Comment by alex-justes** [comment]
So, it just overrides default C and CXX flags. 

Try to add `-DNDEBUG` to  `-DCMAKE_CXX_FLAGS_RELEASE` and to  `-DCMAKE_C_FLAGS_RELEASE` .

**Comment by yurivict** [comment]
IMO, cmake should set ```-DNDEBUG``` based on ```CMAKE_BUILD_TYPE```.

**Comment by arthaud** [comment]
Hi @yurivict,

As @alex-justes and @yurivict said, cmake should use `-DNDEBUG` automatically when using `-DCMAKE_BUILD_TYPE=Release`. This is not handled anywhere in my code.

If you still have the issue, try running `cmake -DCMAKE_BUILD_TYPE=Release [parameters] ..` and `make VERBOSE=1` and post the full logs here. Thank you.

**Comment by yurivict** [comment]
I've just added ```NDEBUG``` to ```CXXFLAGS``` manually, but this should be handled automatically.

**Comment by alex-justes** [comment]
It uses CMAKE_<LANG>_FLAGS_<CONFIGURATION> which were overridden in @yurivict build.

cmake will add NDEBUG automatically if you don't override default flags. Why it should add implicit flags if you set you own flags explicitly? ;) 

P.S.
Sometimes it is needed not to set NDEBUG in Release builds.

**Comment by arthaud** [comment]
Thanks @alex-justes!

@yurivict, since you defined your own `CMAKE_CXX_FLAGS_RELEASE`, you are overriding the default one that contains the `NDEBUG` flag. This is a `cmake` behavior, it's not related to IKOS.

Can I close the issue?

**Comment by yurivict** [comment]
@arthaud ```CMAKE_CXX_FLAGS_RELEASE``` supplied by the system are optimization flags. ```NDEBUG``` is the project-specific flag. ikos cmake scripts should add ```NDEBUG``` based on ```CMAKE_BUILD_TYPE```. The same also happens for many other projects. Systems needs to pass general compilation flags, but project-specific flags are added on top by makefiles.

**Comment by arthaud** [comment]
Sorry @yurivict, I still disagree.

First, `NDEBUG` is not project specific at all. It is used in the C standard library (see `assert.h`), and is also mentioned in the C++ documentation, see https://en.cppreference.com/w/c/error/assert

I shouldn't have to define `NDEBUG` manually, the same way I don't have to set `-g` in debug mode, `-O2` in release mode, etc. This is all the responsibility of CMake.

If you talk to a CMake developer, they will tell you that you shouldn't set any compiler flags directly in your `CMakeLists.txt`, and instead use targets and properties. CMake is an abstraction layer, it should handle all this. For instance, they know better than us when to use `-g` vs `/MTd`, depending on the compiler.

Also, this is the first time I see someone overriding `CMAKE_CXX_FLAGS_RELEASE`. I don't know if this is recommended. Since you are using `-DCMAKE_BUILD_TYPE=Release`, why not just use `CMAKE_CXX_FLAGS`? You will have the same behavior without overriding cmake specific variables. If you want to override cmake generated variables, if have to do it at your own risk (and in this case, add `-DNDEBUG` yourself).

**Comment by yurivict** [comment]
Ok, thanks. I'll close it, because I don't think it's worth having a debate over. It's just a matter of where ```NDEBUG``` is supposed to be defined.

**Comment by arthaud** [comment]
Thank you!

**Comment by arthaud** [comment]
For the record, this was actually fixed in b04b31d60f6bc844274c9ec4eba3a1218ca567bf.

The message `ikos was built in debug mode, the analysis might be slow` now only shows when building with `-DCMAKE_BUILD_TYPE=Debug`, regardless of `-DNDEBUG`.

This way you can compile in release mode but with assertions, which I think is a great idea. We should probably add a specific option for it (similar to `LLVM_ENABLE_ASSERTIONS`).
